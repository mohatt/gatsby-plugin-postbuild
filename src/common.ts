import type { GatsbyNode, Reporter } from 'gatsby'
import Debug from 'debug'

/**
 * Plugin name
 */
export const PLUGIN = 'gatsby-plugin-postbuild'

/**
 * Prints a debug message
 */
export const debug = (message: string, ...params: any[]): void => {
  createDebug()(message, ...params)
}

/**
 * Creates a debug instance for the given namespace
 */
export const createDebug = (namespace = ''): Function => {
  let ns = PLUGIN
  if (namespace !== '') {
    ns += ':' + namespace
  }
  const debug = Debug(ns)
  return (message: string, ...params: any[]) => {
    message = params.length > 0
      ? `${message} %O \n\n`
      : `${message} \n\n`
    debug(message, ...params)
  }
}

/**
 * Manages Gatsby's reporter instance, providing functionalities for structured error logging,
 * warnings, and instance management.
 *
 * @internal
 */
export const reporter = (() => {
  let ref: Reporter | undefined

  // Initializes reporter
  /**
   * Initializes and caches the Gatsby reporter instance.
   *
   * If `setErrorMap` is available, it registers custom error messages
   * for better debugging in Gatsby's CLI.
   */
  const initialize = (instance: Reporter): Reporter => {
    instance.setErrorMap({
      ['1400' satisfies PluginErrorMeta['id']]: {
        text: (context: PluginErrorMeta['context']) => context.message,
        category: 'THIRD_PARTY',
        level: 'ERROR',
        type: 'PLUGIN',
      },
    })

    ref = instance
    return instance
  }

  /**
   * Retrieves the cached reporter instance.
   */
  const get = (): Reporter => {
    return ref
  }

  const createError = (api: keyof GatsbyNode, err: string | PluginError | Error): PluginErrorMeta => {
    const prefix = `The plugin threw an error during "${api}" hook`

    let title: string | undefined
    let mainError: Error | undefined

    if (err instanceof PluginError) {
      title = err.message
      mainError = err.originalError
    } else if (err instanceof Error) {
      mainError = err
    } else {
      title = err
    }

    return {
      id: '1400',
      context: {
        message: mainError && title ? `${prefix}:\n ${title}` : prefix,
      },
      error: mainError ?? (title ? new Error(title) : undefined),
    }
  }

  /**
   * Logs an error and terminates the build process.
   *
   * - Uses Gatsby's `panic()` method to log structured errors.
   * - If an `Error` object is provided, includes it in the error report.
   * - Otherwise, creates a new error instance from the given message.
   *
   * @param api The Gatsby Node API that caused the error.
   * @param err The error message or PluginError object to display.
   *
   * @throws - This function never returns; it always terminates execution.
   */
  const error = (api: keyof GatsbyNode, err: string | PluginError | Error): never => {
    const meta = createError(api, err)
    if (!ref) {
      throw err
    }
    return ref.panic(meta)
  }

  /**
   * Logs a warning message in the Gatsby CLI.
   *
   * @param message The warning message to display.
   */
  const warning = (message: string): void => {
    const prefix = `"${PLUGIN}" might not be working properly`
    const warning = `${prefix}:\n ${message}`
    if (!ref) {
      throw new Error(warning)
    }
    ref.warn(warning)
  }

  return { initialize, get, error, createError, warning }
})()

/**
 * Wraps a Gatsby API function with error handling.
 *
 * @param api - The name of the Gatsby API being wrapped (e.g., `"onCreatePage"`).
 * @param fn - The original Gatsby API function to be wrapped.
 *
 * @returns The wrapped function with error handling.
 */
export const createPluginExport = <T extends keyof GatsbyNode, F extends GatsbyNode[T]>(
  api: T,
  fn: F,
): F => {
  return ((...args: any[]) => {
    try {
      // Call the original function with provided arguments
      const result = (fn as any)(...args)

      // If the function returns a Promise, attach error handling
      if (result instanceof Promise) {
        return result.catch((error) => reporter.error(api, error))
      }

      // Return the result (for synchronous functions)
      return result
    } catch (error) {
      // Catch synchronous errors and report them
      reporter.error(api, error)
    }
  }) as F
}

/**
 * The main plugin error representation in Gatsby's error map.
 */
export interface PluginErrorMeta {
  id: '1400'
  context: {
    message: string
  }
  error: Error
}

/**
 * Custom error class for plugin-related errors.
 */
export class PluginError extends Error {
  /**
   * @param message - A description of the error.
   * @param originalError - The original error that caused this issue (optional).
   */
  constructor(
    message: string,
    public readonly originalError?: Error,
  ) {
    super(message)
  }
}
