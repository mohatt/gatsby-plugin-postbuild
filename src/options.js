export const defaults = {
  enabled: true,
  report: true,
  reportConsole: true,
  purgecss: {
    allowSymbols: false,
    rejected: true,
    fontFace: false,
    keyframes: false,
    variables: false
  }
}

export function schema (Joi) {
  return Joi.object({
    enabled: Joi.boolean()
      .description('Whether to run the purgecss task or not.'),
    report: Joi.boolean()
      .description('Write a "postbuild.log.json" file in "public" with all the changes made.'),
    reportConsole: Joi.boolean()
      .description('Print a summary report during build with all the changes made.'),
    purgecss: Joi.object({
      allowSymbols: Joi.boolean()
        .description('Allow CSS selectors to contain symbols (e.g TailwindCSS).'),
      rejected: Joi.boolean()
        .description('Write a log file in "public" with the rejected selectors for every file.'),
      defaultExtractor: Joi.function()
        .description('A custom PurgeCSS extractor to be used instead of the default one.'),
      safelist: Joi.alternatives().try(
        Joi.object({}).unknown(true),
        Joi.array().items(Joi.string())
      ).description('Selectors that are safe to leave in the final CSS.'),
      blocklist: Joi.array().items(Joi.string())
        .description('Selectors that are blocked from appearing in the final CSS.'),
      fontFace: Joi.boolean()
        .description('Remove any unused @font-face rules in your css.'),
      keyframes: Joi.boolean()
        .description('Remove any unused @keyframes rules in your css.'),
      variables: Joi.boolean()
        .description('Remove any unused variables in your css.')
    })
  })
}
