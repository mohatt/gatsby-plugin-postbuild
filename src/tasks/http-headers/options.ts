import type { ITaskApiOptions, ITaskOptions } from '@postbuild'
import { PROVIDER_TYPES, ProviderType } from './lib/providers'
import type Link from './lib/link'
import type Meta from './lib/meta'

/**
 * Header value is either a string for single-value headers or
 *  an array of strings for multi-value headers
 *
 *  Example of a single-value header:
 *   {
 *     'cache-control': 'public, max-age=31536000, immutable'
 *   }
 *
 *  Examples of a multi-value header:
 *   {
 *    'link': [
 *      '</script.js>; rel=preload; as=script',
 *      '</image.png>; rel=preload; as=image'
 *    ]
 *   }
 */
export interface IHeadersMap {
  [name: string]: string | string[]
}

/**
 * Map of paths to path headers
 */
export interface IPathHeadersMap {
  [path: string]: IHeadersMap
}

/**
 * Task options interface
 */
export interface IHttpHeadersTaskOptions extends ITaskOptions {
  provider: ProviderType
  headers: IPathHeadersMap
  security: boolean
  caching: boolean
  transformPathLinks: (links: Link[], path: string) => Link[]
  removeLinkTags: boolean
  transformPathMetas: (metas: Meta[], path: string) => Meta[]
  removeMetaTags: boolean
}

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IHttpHeadersTaskOptions> = {
  defaults: {
    enabled: false,
    ignore: [],
    provider: 'netlify',
    headers: {},
    security: true,
    caching: true,
    transformPathLinks: (links) => links,
    removeLinkTags: true,
    transformPathMetas: (metas) => metas,
    removeMetaTags: true,
  },
  schema: (joi) => {
    const nonEmptyString = joi.string().trim().required()
    return joi.object({
      provider: joi
        .string()
        .valid(...PROVIDER_TYPES)
        .description('Headers file provider.'),
      headers: joi
        .object()
        .pattern(
          nonEmptyString,
          joi
            .object()
            .pattern(
              joi.string().pattern(/^[\w-]+$/),
              joi.alternatives().try(nonEmptyString, joi.array().items(nonEmptyString)),
            ),
        )
        .description('Headers to merge with the generated headers.'),
      security: joi.boolean().description('Adds useful security headers.'),
      caching: joi.boolean().description('Adds useful caching headers to immutable asset paths.'),
      transformPathLinks: joi
        .function()
        .maxArity(2)
        .description('Callback for manipulating links under each path.'),
      removeLinkTags: joi
        .boolean()
        .description('Removes the HTML link tags processed by the task.'),
      transformPathMetas: joi
        .function()
        .maxArity(2)
        .description('Callback for manipulating metas under each path.'),
      removeMetaTags: joi
        .boolean()
        .description('Removes the HTML meta tags processed by the task.'),
    })
  },
}

declare module 'gatsby-plugin-postbuild' {
  interface IOptions {
    'http-headers': IHttpHeadersTaskOptions
  }
}
