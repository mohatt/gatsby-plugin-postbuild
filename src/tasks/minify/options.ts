import type { MinifyOptions as TerserOptions } from 'terser'
import type { ITaskApiOptions, ITaskOptions } from '@postbuild'

/**
 * Task options interface
 */
export interface IMinifyTaskOptions extends ITaskOptions {
  script: TerserOptions | boolean
  style: [string, object] | object | Function | boolean
}

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IMinifyTaskOptions> = {
  defaults: {
    enabled: false,
    ignore: [],
    script: {},
    style: ['default', {}],
  },
  schema: (joi) => {
    return joi.object({
      script: joi
        .alternatives()
        .try(joi.boolean(), joi.object().unknown(true))
        .description('Options passed to javascript minifier.'),
      style: joi
        .alternatives()
        .try(
          joi.boolean(),
          joi.object(),
          joi.array().items(joi.string(), joi.object().unknown(true)),
          joi.function().maxArity(0),
        )
        .description('Options passed to css minifier.'),
    })
  },
}

declare module 'gatsby-plugin-postbuild' {
  interface IOptions {
    minify: IMinifyTaskOptions
  }
}
