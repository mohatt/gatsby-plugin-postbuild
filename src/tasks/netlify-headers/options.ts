import type { ITaskApiOptions, ITaskOptions } from '~/tasks'

/**
 * Task options interface
 */
export type IOptions = ITaskOptions & {
  headers: {
    [path: string]: string[]
  }
  security: boolean
  caching: boolean
  cachingAssetTypes: string[]
  removeLinkTags: boolean
}

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IOptions> = {
  defaults: {
    enabled: true,
    ignore: [],
    headers: {},
    security: true,
    caching: true,
    cachingAssetTypes: ['image', 'script', 'style', 'font'],
    removeLinkTags: true
  },
  // @todo write desc
  schema: (joi) => {
    return joi.object({
      headers: joi.object()
        .pattern(joi.string(), joi.array().items(joi.string()))
        .description('@todo'),
      security: joi.boolean()
        .description('@todo'),
      caching: joi.boolean()
        .description('@todo'),
      cachingAssetTypes: joi.array()
        .items(joi.string())
        .description('@todo'),
      removeLinkTags: joi.boolean()
        .description('@todo')
    })
  }
}
