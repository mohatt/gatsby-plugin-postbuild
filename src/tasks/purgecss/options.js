export const defaults = {
  // Options consumend by the task
  enabled: true,
  report: true,
  reportConsole: true,
  tailwind: false,
  // Options consumend by both the task and PurgeCSS
  rejected: true
  // Options consumend by PurgeCss
  // We leave that to PurgeCss
}

export function schema (Joi) {
  return Joi.object({
    enabled: Joi.boolean()
      .description('Whether to run the purgecss task or not.'),
    report: Joi.boolean()
      .description('Write a "purgecss.log.json" file in "public" with all the changes made.'),
    reportConsole: Joi.boolean()
      .description('Print a summary report during build with all the changes made.'),
    tailwind: Joi.boolean()
      .description('Use custom extractor for tailwind.'),
    rejected: Joi.boolean()
      .description('Write a log file in "public" with the rejected selectors for every file.'),
    defaultExtractor: Joi.function()
      .description('A custom PurgeCSS extractor to be used instead of the default one.'),
    extractors: Joi.array().items(Joi.object({}).unknown(true))
      .description('A list of custom PurgeCSS extractors to be used for certain file types.'),
    safelist: Joi.alternatives().try(
      Joi.object({}).unknown(true),
      Joi.array().items(Joi.string())
    ).description('Selectors that are safe to leave in the final CSS.'),
    blocklist: Joi.array().items(Joi.string())
      .description('Selectors that are blocked from appearing in the final CSS.'),
    fontFace: Joi.boolean()
      .description('Remove any unused @font-face rules in your css.'),
    keyframes: Joi.boolean()
      .description('Remove any unused @keyframes rules in your css.'),
    variables: Joi.boolean()
      .description('Remove any unused variables in your css.')
  })
}
