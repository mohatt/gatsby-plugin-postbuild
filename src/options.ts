import { GatsbyJoi } from './gatsby'
import { defaultOptions } from 'purgecss'
export type PurgeCSSOptions = typeof defaultOptions

/**
 * Plugin options interface
 */
export interface IPluginOptions {
  enabled: boolean
  report: boolean
  reportConsole: boolean
  allowSymbols: boolean
  ignoreFiles: {
    webpack: string[]
    pages: string[]
    css: string[]
    js: string[]
  }
  purgecss: Pick<PurgeCSSOptions, 'defaultExtractor'|'rejected'|'fontFace'|'keyframes'|'variables'|'blocklist'|'safelist'>
}

/**
 * Default values for plugin options
 *  we are merging purgecss defaults as well while
 *  overridding some values
 */
export const defaults: IPluginOptions = {
  enabled: true,
  report: true,
  reportConsole: true,
  ignoreFiles: {
    webpack: ['app', 'polyfill'],
    pages: [],
    css: [],
    js: []
  },
  allowSymbols: false,
  purgecss: {
    ...defaultOptions,
    rejected: true
  }
}

/**
 * Plugin options schema
 */
export function schema (Joi: GatsbyJoi): GatsbyJoi {
  return Joi.object({
    enabled: Joi.boolean()
      .description('Whether to run the purgecss task or not.'),
    report: Joi.boolean()
      .description('Write a "postbuild.log.json" file in "/public" with all the changes made.'),
    reportConsole: Joi.boolean()
      .description('Print a summary report during build with all the changes made.'),
    ignoreFiles: Joi.object({
      webpack: Joi.array().items(Joi.string())
        .description('Webpack chunck names to ignore while purging CSS.'),
      pages: Joi.array().items(Joi.string())
        .description('Pages to exclude from optimiztion.'),
      css: Joi.array().items(Joi.string())
        .description('Webpack chunck names to ignore while purging CSS.'),
      js: Joi.array().items(Joi.string())
        .description('Webpack chunck names to ignore while purging CSS.')
    }).description('CSS/JavaScript files to ignore during optimization. File names should be relative to "/public" directory'),
    allowSymbols: Joi.boolean()
      .description('Sets a custom PurgeCSS extractor that allows CSS selectors to contain symbols (e.g TailwindCSS).'),
    purgecss: Joi.object({
      rejected: Joi.boolean()
        .description('Write a log file in "/public" with the rejected selectors for every file.'),
      defaultExtractor: Joi.function()
        .description('A custom PurgeCSS extractor to be used instead of the default one.'),
      safelist: Joi.alternatives().try(
        Joi.object({}).unknown(true),
        Joi.array().items(Joi.string())
      ).description('Selectors that are safe to leave in the final CSS.'),
      blocklist: Joi.array().items(Joi.string())
        .description('Selectors that are blocked from appearing in the final CSS.'),
      fontFace: Joi.boolean()
        .description('Remove any unused @font-face rules in your css.'),
      keyframes: Joi.boolean()
        .description('Remove any unused @keyframes rules in your css.'),
      variables: Joi.boolean()
        .description('Remove any unused variables in your css.')
    })
  })
}
