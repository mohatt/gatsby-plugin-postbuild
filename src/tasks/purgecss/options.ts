import _ from 'lodash'
import { defaultOptions as purgecssDefaults } from 'purgecss'
import type { ITaskApiOptions, ITaskOptions } from '~/tasks'
type PurgecssDefaults = typeof purgecssDefaults

/**
 * Extracts union type from array values
 */
type ArrayUnion<T extends readonly unknown[]> = T extends ReadonlyArray<infer A> ? A : never

/**
 * Options imported from PurgeCss
 */
export const purgecssImportedOptions = [
  'defaultExtractor',
  'safelist',
  'blocklist',
  'fontFace',
  'keyframes',
  'variables'
] as const

/**
 * Task options interface
 */
export type IOptions = ITaskOptions & {
  reportRejected: boolean
  allowSymbols: boolean
  ignoreAssets: {
    webpack: string[]
    css: string[]
    js: string[]
  }
} & Pick<PurgecssDefaults, ArrayUnion<typeof purgecssImportedOptions>>

/**
 * All options accepted by purgecss lib
 */
export type PurgecssOptions = Omit<Partial<PurgecssDefaults>, 'content'|'css'> & Pick<PurgecssDefaults, 'content'|'css'>

/**
 * Options API exports
 */
export const options: ITaskApiOptions<IOptions> = {
  defaults: {
    enabled: true,
    ignore: [],
    ignoreAssets: {
      webpack: ['app', 'polyfill'],
      css: [],
      js: []
    },
    reportRejected: true,
    allowSymbols: false,
    ..._.pick(purgecssDefaults, purgecssImportedOptions)
  },
  schema: (joi) => {
    return joi.object({
      reportRejected: joi.boolean()
        .description('Write a log file in `/public` with the rejected selectors for every file.'),
      ignoreAssets: joi.object({
        webpack: joi.array().items(joi.string())
          .description('Webpack chunck names to ignore while purging CSS.'),
        css: joi.array().items(joi.string())
          .description('CSS files to exclude from processing.'),
        js: joi.array().items(joi.string())
          .description('JavaScript files to ignore while purging CSS.')
      }).description('CSS/JavaScript files to ignore during optimization. File names should be relative to `/public` directory'),
      allowSymbols: joi.boolean()
        .description('Sets a custom PurgeCSS extractor that allows CSS selectors to contain symbols (e.g TailwindCSS).'),
      defaultExtractor: joi.function()
        .description('A custom PurgeCSS extractor to be used instead of the default one.'),
      safelist: joi.alternatives().try(
        joi.object({}).unknown(true),
        joi.array().items(joi.string())
      ).description('Selectors that are safe to leave in the final CSS.'),
      blocklist: joi.array().items(joi.string())
        .description('Selectors that are blocked from appearing in the final CSS.'),
      fontFace: joi.boolean()
        .description('Remove any unused `@font-face` rules in your css.'),
      keyframes: joi.boolean()
        .description('Remove any unused `@keyframes` rules in your css.'),
      variables: joi.boolean()
        .description('Remove any unused variables in your css.')
    })
  }
}
