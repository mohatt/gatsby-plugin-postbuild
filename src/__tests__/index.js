import { testPluginOptionsSchema } from 'gatsby-plugin-utils'
import { pluginOptionsSchema, onPreBootstrap, onPostBuild } from '~/index'
import Postbuild from '~/postbuild'
import { schema } from '~/options'
import { PostbuildError } from '~/common'
import optionsFixtures from '#/__fixtures__/options'

jest.mock('~/postbuild')
const gatsbyActivity = {
  start: jest.fn(),
  setStatus: jest.fn(),
  panic: jest.fn(),
  end: jest.fn()
}
const gatsby = {
  reporter: {
    panic: jest.fn(),
    activityTimer: jest.fn().mockImplementation(() => gatsbyActivity)
  },
  tracing: {
    parentSpan: 'foo'
  }
}

describe('pluginOptionsSchema', () => {
  const postbuild = Postbuild.mock.instances[0]
  test('runs correctly', () => {
    expect(() => pluginOptionsSchema({})).not.toThrow()
    postbuild.getOptionsSchemas = () => { throw new Error('foo') }
    expect(() => pluginOptionsSchema({})).toThrowErrorMatchingSnapshot()
  })

  describe('validates user options against schema', () => {
    for (const { title, options } of optionsFixtures) {
      test(title, async () => {
        postbuild.getOptionsSchemas = joi => schema(joi)
        const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options)
        expect(isValid).toMatchSnapshot()
        expect(errors).toMatchSnapshot()
      })
    }
  })
})

describe('onPreBootstrap', () => {
  const postbuild = Postbuild.mock.instances[0]
  test('runs correctly', async () => {
    const args = [gatsby, {}]
    await expect(onPreBootstrap(...args)).resolves.toBeUndefined()
    postbuild.bootstrap = () => {
      throw new Error('foo')
    }
    onPreBootstrap(...args)
    postbuild.bootstrap = () => {
      throw new PostbuildError('something wrong', new TypeError('foo'))
    }
    onPreBootstrap(...args)
    expect(gatsby.reporter.panic.mock.calls).toMatchSnapshot()
  })
})

describe('onPostBuild', () => {
  const postbuild = Postbuild.mock.instances[0]
  test('runs correctly', async () => {
    await expect(onPostBuild(gatsby)).resolves.toBeUndefined()
    postbuild.run = () => {
      throw new Error('foo')
    }
    await onPostBuild(gatsby)
    postbuild.run = () => {
      throw new PostbuildError('something wrong', new TypeError('foo'))
    }
    await onPostBuild(gatsby)
    expect(gatsby.reporter.activityTimer.mock.calls).toMatchSnapshot()
    expect(gatsbyActivity.start.mock.calls).toMatchSnapshot()
    expect(gatsbyActivity.panic.mock.calls).toMatchSnapshot()
    expect(gatsbyActivity.end.mock.calls).toMatchSnapshot()
  })
})
