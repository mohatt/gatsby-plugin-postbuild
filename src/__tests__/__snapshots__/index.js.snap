// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onPostBuild runs correctly 1`] = `
Array [
  Array [
    "gatsby-plugin-postbuild",
    Object {
      "parentSpan": "foo",
    },
  ],
  Array [
    "gatsby-plugin-postbuild",
    Object {
      "parentSpan": "foo",
    },
  ],
  Array [
    "gatsby-plugin-postbuild",
    Object {
      "parentSpan": "foo",
    },
  ],
]
`;

exports[`onPostBuild runs correctly 2`] = `
Array [
  Array [],
  Array [],
  Array [],
]
`;

exports[`onPostBuild runs correctly 3`] = `
Array [
  Array [
    Object {
      "context": [Error: foo],
      "error": [Error: foo],
      "id": "gatsby-plugin-postbuild_10000",
    },
  ],
  Array [
    Object {
      "context": Object {
        "message": "something wrong",
      },
      "error": [TypeError: foo],
      "id": "gatsby-plugin-postbuild_10000",
    },
  ],
]
`;

exports[`onPostBuild runs correctly 4`] = `
Array [
  Array [],
  Array [],
  Array [],
]
`;

exports[`onPreBootstrap runs correctly 1`] = `
Array [
  Array [
    Object {
      "context": [Error: foo],
      "error": [Error: foo],
      "id": "gatsby-plugin-postbuild_10000",
    },
  ],
  Array [
    Object {
      "context": Object {
        "message": "something wrong",
      },
      "error": [TypeError: foo],
      "id": "gatsby-plugin-postbuild_10000",
    },
  ],
]
`;

exports[`pluginOptionsSchema runs correctly 1`] = `"Error occured while initializing \\"gatsby-plugin-postbuild\\" plugin: foo"`;

exports[`pluginOptionsSchema validates user options against schema should invalidate empty options 1`] = `false`;

exports[`pluginOptionsSchema validates user options against schema should invalidate empty options 2`] = `
Array [
  "\\"ignore[0]\\" is not allowed to be empty",
  "\\"extensions.txt.concurrency\\" must be greater than or equal to 1",
  "\\"extensions.txt.strategy\\" must be one of [steps, parallel]",
  "\\"extensions.txt.strategy\\" is not allowed to be empty",
  "\\"defaultStrategy\\" must be one of [steps, parallel]",
  "\\"defaultStrategy\\" is not allowed to be empty",
  "\\"defaultConcurrency\\" must be greater than or equal to 1",
]
`;

exports[`pluginOptionsSchema validates user options against schema should invalidate invalid options 1`] = `false`;

exports[`pluginOptionsSchema validates user options against schema should invalidate invalid options 2`] = `
Array [
  "\\"events.on.load\\" is not allowed",
  "\\"events.html.content\\" is not allowed",
  "\\"events.txt.parse\\" is not allowed",
  "\\"extensions.txt.concurrency\\" must be greater than or equal to 1",
  "\\"extensions.txt.strategy\\" must be one of [steps, parallel]",
  "\\"defaultStrategy\\" must be one of [steps, parallel]",
  "\\"defaultConcurrency\\" must be greater than or equal to 1",
]
`;

exports[`pluginOptionsSchema validates user options against schema should invalidate options with incorrect data types 1`] = `false`;

exports[`pluginOptionsSchema validates user options against schema should invalidate options with incorrect data types 2`] = `
Array [
  "\\"enabled\\" must be a boolean",
  "\\"report\\" must be a boolean",
  "\\"ignore\\" must be an array",
  "\\"events.on.bootstrap\\" must be of type function",
  "\\"events.html.parse\\" must be of type function",
  "\\"events.foo\\" must be of type object",
  "\\"events.txt.content\\" must be of type function",
  "\\"extensions.foo\\" must be of type object",
  "\\"extensions.txt.concurrency\\" must be a number",
  "\\"extensions.txt.strategy\\" must be one of [steps, parallel]",
  "\\"extensions.txt.strategy\\" must be a string",
  "\\"defaultStrategy\\" must be one of [steps, parallel]",
  "\\"defaultStrategy\\" must be a string",
  "\\"defaultConcurrency\\" must be a number",
]
`;

exports[`pluginOptionsSchema validates user options against schema should invalidate unknown options 1`] = `false`;

exports[`pluginOptionsSchema validates user options against schema should invalidate unknown options 2`] = `
Array [
  "\\"unknown\\" is not allowed",
  "\\"foo\\" is not allowed",
]
`;

exports[`pluginOptionsSchema validates user options against schema should validate correct options 1`] = `true`;

exports[`pluginOptionsSchema validates user options against schema should validate correct options 2`] = `Array []`;
